### PR TITLE
Fix bug with value distribution by asset not respecting ETH=ETH2 setting

### LINKED ISSUE #4578: Value distribution by asset when eth == eth2
## Problem Definition

When ETH2 is treated like ETH the value distribution by asset graph is off and is not considering them to be the same asset properly as seen in the screenshot.

## Task

Fix it. It should be easy to do so in the backend.

### PR DESCRIPTION
**Pull request by prettyirrelevant**
Closes #4578

### CODE REVIEW

File: rotkehlchen/db/dbhandler.py [hunk 1]
```diff
@@ -2768,27 +2768,49 @@ def get_latest_asset_value_distribution(self) -> List[DBAssetBalance]:
...
+            treat_eth2_as_eth = self.get_settings(cursor).treat_eth2_as_eth
             cursor.execute(
                 'SELECT time, currency, amount, usd_value, category FROM timed_balances WHERE '
                 'time=(SELECT MAX(time) from timed_balances) AND category = ? ORDER BY '
                 'CAST(usd_value AS REAL) DESC;',
                 (BalanceType.ASSET.serialize_for_db(),),  # pylint: disable=no-member
             )
             asset_balances = []
+            eth_balance: Dict[str, str] = {
+                'time': '0',
```
Comment (yabirgb) [inline rotkehlchen/db/dbhandler.py:2780]: Is there any reason to treat this as string instead of integer? In the database and everywhere else we use integers for timestamps I believe

File: rotkehlchen/db/dbhandler.py [hunk 2]
```diff
@@ -2768,27 +2768,49 @@ def get_latest_asset_value_distribution(self) -> List[DBAssetBalance]:
...
                     continue
+                # show eth & eth2 as eth in value distribution by asset
+                if treat_eth2_as_eth is True and asset in (A_ETH, A_ETH2):
+                    eth_balance['time'] = str(result[0])
+                    eth_balance['amount'] = str(FVal(eth_balance['amount']) + FVal(result[2]))
+                    eth_balance['usd_value'] = str(FVal(eth_balance['usd_value']) + FVal(result[3]))  # noqa: E501
+                else:
+                    asset_balances.append(
+                        DBAssetBalance(
+                            time=result[0],
```
Comment (yabirgb) [inline rotkehlchen/db/dbhandler.py:2796]: isn't the linter complaining here about missing a casting to `Timestamp`?

File: rotkehlchen/db/dbhandler.py [hunk 3]
```diff
@@ -2768,27 +2768,49 @@ def get_latest_asset_value_distribution(self) -> List[DBAssetBalance]:
...
-                        time=result[0],
-                        asset=asset,
-                        amount=result[2],
-                        usd_value=result[3],
-                        category=BalanceType.deserialize_from_db(result[4]),
+                        time=Timestamp(int(eth_balance['time'])),
+                        asset=A_ETH,
+                        amount=eth_balance['amount'],
+                        usd_value=eth_balance['usd_value'],
+                        category=BalanceType.ASSET,
```
Comment (yabirgb) [inline rotkehlchen/db/dbhandler.py:2811]: Also you are duplicating code here. Maybe you can save all the info in some kind of list and the create the `DBAssetBalance` from that clean list avoiding duplication

**Review by yabirgb — CHANGES_REQUESTED** [review]
some comments

File: rotkehlchen/db/dbhandler.py (hunk 2 above)
Comment (prettyirrelevant) [reply rotkehlchen/db/dbhandler.py:2796]: It has a type of `Any`

File: rotkehlchen/db/dbhandler.py [hunk 4]
```diff
@@ -2768,27 +2768,49 @@ def get_latest_asset_value_distribution(self) -> List[DBAssetBalance]:
...
+            }
             for result in cursor:
                 asset = Asset(result[1])
                 if asset in ignored_assets:
                     continue
+                # show eth & eth2 as eth in value distribution by asset
+                if treat_eth2_as_eth is True and asset in (A_ETH, A_ETH2):
+                    eth_balance['time'] = str(result[0])
+                    eth_balance['amount'] = str(FVal(eth_balance['amount']) + FVal(result[2]))
+                    eth_balance['usd_value'] = str(FVal(eth_balance['usd_value']) + FVal(result[3]))  # noqa: E501
```
[the remark this answers is not in the document]
Comment (prettyirrelevant) [reply rotkehlchen/db/dbhandler.py:2792]: The time is the same for every entry due to the SQL query.
```sql
SELECT time, currency, amount, usd_value, category FROM timed_balances WHERE '
'time=(SELECT MAX(time) from timed_balances) AND category = ? ORDER BY '
'CAST(usd_value AS REAL) DESC;
  ```

[the remark this answers is not in the document]
Comment (yabirgb) [reply rotkehlchen/db/dbhandler.py:2792]: Okey that is fine then

File: rotkehlchen/db/dbhandler.py (hunk 3 above)
Comment (prettyirrelevant) [reply rotkehlchen/db/dbhandler.py:2811]: I don't totally get you here

Comment (yabirgb) [reply rotkehlchen/db/dbhandler.py:2811]: I was thinking in how to remove the duplication of the object creation but it would mean to go through the data twice... Does any other good idea come to your mind? otherwise I guess is fine

Comment (prettyirrelevant) [reply rotkehlchen/db/dbhandler.py]: Also, it's not duplication as I'm only adding the ETH balance after the loop. Unless I don't get you

**Review by yabirgb — COMMENTED** [review]
Looks okey but in my account where I have ETH >>> ETH2 the UI was showing the ETH portion with the key "Others" since ETH2 was before in the "Other assets" section of the pie chart. Is that something that needs adjustment in the frontend @kelsos @lukicenturi ?

File: rotkehlchen/db/dbhandler.py [hunk 5]
```diff
@@ -2768,27 +2768,49 @@ def get_latest_asset_value_distribution(self) -> List[DBAssetBalance]:
...
+                'usd_value': '0',
+            }
             for result in cursor:
                 asset = Asset(result[1])
                 if asset in ignored_assets:
                     continue
+                # show eth & eth2 as eth in value distribution by asset
+                if treat_eth2_as_eth is True and asset in (A_ETH, A_ETH2):
+                    eth_balance['time'] = int(result[0])
+                    eth_balance['amount'] = str(FVal(eth_balance['amount']) + FVal(result[2]))
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2791]: Again here just like I asked you the last time you touched some code ilke this since now `DBAssetBalance` is being used in calculations it should have FVals and should be properly serialized and deserialized just like the change you did with `SingleDBAssetBalance`.

There is no point in doing what you are doing here ... you are turning string to FVal, add them and then back to strings. This means that the data structure needs to change.

**Comment by LefterisJP** [comment]
> Looks okey but in my account where I have ETH >>> ETH2 the UI was showing the ETH portion with the key "Others" since ETH2 was before in the "Other assets" section of the pie chart. Is that something that needs adjustment in the frontend @kelsos @lukicenturi ?

What do you mean?

**Review by LefterisJP — COMMENTED** [review]
Will also need a rebase

**Comment by yabirgb** [comment]
> > Looks okey but in my account where I have ETH >>> ETH2 the UI was showing the ETH portion with the key "Others" since ETH2 was before in the "Other assets" section of the pie chart. Is that something that needs adjustment in the frontend @kelsos @lukicenturi ?
> 
> What do you mean?

Tested it in the frontend and the distribution section instead of having the name as `ETH` for ETH + ETH2 it had `others`

**Comment by LefterisJP** [comment]
> > > Looks okey but in my account where I have ETH >>> ETH2 the UI was showing the ETH portion with the key "Others" since ETH2 was before in the "Other assets" section of the pie chart. Is that something that needs adjustment in the frontend @kelsos @lukicenturi ?
> > 
> > 
> > What do you mean?
> 
> Tested it in the frontend and the distribution section instead of having the name as `ETH` for ETH + ETH2 it had `others`

No that is not a frontend problem.

This is something else @prettyirrelevant needs to solve. The DB query originally returns them in `CAST(usd_value AS REAL) DESC;` descending value.

What @prettyirrelevant destroys this, and as such it breaks the feature. If `ETH==ETH2` it's always added as the last, so it will always be in "Others".

So @prettyirrelevant what you need to do is either move the whole logic in the DB query, or re-sort the list by descending balance or just iterate it and put ETH/ETH2 at the right place. The last option is probably better than sorting again since the list is already sorted.

File: rotkehlchen/db/dbhandler.py [hunk 6]
```diff
@@ -2772,27 +2772,54 @@ def get_latest_asset_value_distribution(self) -> List[DBAssetBalance]:
...
+                # respect descending order `usd_value`
+                for index, balance in enumerate(asset_balances):
+                    if FVal(eth_balance['usd_value']) > balance.usd_value:
+                        asset_balances.insert(
+                            index,
+                            DBAssetBalance(
+                                time=Timestamp(FVal(eth_balance['time']).to_int(True)),
+                                asset=A_ETH,
+                                amount=FVal(eth_balance['amount']),
+                                usd_value=FVal(eth_balance['usd_value']),
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2818]:
```suggestion
                                usd_value=eth_balance['usd_value'],
```

File: rotkehlchen/db/dbhandler.py [hunk 7]
```diff
@@ -2772,27 +2772,54 @@ def get_latest_asset_value_distribution(self) -> List[DBAssetBalance]:
...
+                        ),
+                    )
+            # only add the eth_balance if it contains a balance > 0
+            if FVal(eth_balance['amount']) > ZERO or FVal(eth_balance['usd_value']) > ZERO:
+                # respect descending order `usd_value`
+                for index, balance in enumerate(asset_balances):
+                    if FVal(eth_balance['usd_value']) > balance.usd_value:
+                        asset_balances.insert(
+                            index,
+                            DBAssetBalance(
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2814]: if you initialize `eth_balance` as a `DBAssetBalance` above this entire function is easier.

File: rotkehlchen/db/snapshots.py [hunk 8]
```diff
@@ -54,9 +54,9 @@ def get_timed_balances(
                     DBAssetBalance(
                         category=BalanceType.deserialize_from_db(data[0]),
                         time=data[1],
-                        amount=data[2],
+                        amount=FVal(data[2]),
```
Comment (LefterisJP) [inline rotkehlchen/db/snapshots.py:57]: It would make sense to create a `DBAssetBalance.deserialize_from_db()` function and use it here

File: rotkehlchen/db/utils.py [hunk 9]
```diff
@@ -151,27 +151,27 @@ class DBAssetBalance(NamedTuple):
     category: BalanceType
     time: Timestamp
     asset: Asset
-    amount: str
-    usd_value: str
+    amount: FVal
```
Comment (LefterisJP) [inline rotkehlchen/db/utils.py:154]: you can also turn this into a data class if you need it at any point to be mutable.

File: rotkehlchen/db/utils.py [hunk 10]
```diff
@@ -151,27 +151,27 @@ class DBAssetBalance(NamedTuple):
...
+    ) -> Dict[str, Union[str, int, FVal]]:
         if export_data:
             return {
                 'timestamp': timestamp_to_date(self.time, '%Y-%m-%d %H:%M:%S'),
                 'category': self.category.serialize(),
                 'asset': str(self.asset),
-                'amount': self.amount,
-                f'{export_data[0].symbol.lower()}_value': str(FVal(self.usd_value) * export_data[1]),  # noqa: 501
+                'amount': str(self.amount),
+                f'{export_data[0].symbol.lower()}_value': FVal(self.usd_value) * export_data[1],
```
Comment (LefterisJP) [inline rotkehlchen/db/utils.py:167]: why did you remove the string constructor from here? Isn't this to serialize? This way you will end up with an FVal here.

File: rotkehlchen/db/utils.py [hunk 11]
```diff
@@ -151,27 +151,27 @@ class DBAssetBalance(NamedTuple):
...
-    amount: str
-    usd_value: str
+    amount: FVal
+    usd_value: FVal
 
     def serialize(
             self,
             export_data: Optional[Tuple[Asset, Price]] = None,
-    ) -> Dict[str, Union[str, int]]:
+    ) -> Dict[str, Union[str, int, FVal]]:
```
Comment (LefterisJP) [inline rotkehlchen/db/utils.py:160]: Please add a docstring to this function. I have had a hard time to understand what `export_data` is and what it does.

File: rotkehlchen/rotkehlchen.py [hunk 12]
```diff
@@ -832,7 +832,6 @@ def query_balances(
                     problem_free=problem_free,
                     save_despite_errors=save_despite_errors,
                 )
-
```
Comment (LefterisJP) [inline rotkehlchen/rotkehlchen.py:835]: Please don't randomly remove lines in other files in PRs. Leave it as it was.

File: rotkehlchen/tests/api/test_statistics.py [hunk 13]
```diff
@@ -303,19 +313,35 @@ def assert_okay_by_location(response):
...
+                'ETH': get_asset_balance_total(A_ETH, setup) + get_asset_balance_total(A_ETH2, setup),  # noqa: E501
+                'BTC': get_asset_balance_total(A_BTC, setup),
+                'EUR': get_asset_balance_total(A_EUR, setup),
+                A_RDN.identifier: get_asset_balance_total(A_RDN, setup),
+            }
+            for entry in result:
+                assert len(entry) == 5
+                assert entry['time'] >= start_time
+                assert entry['category'] == 'asset'
+                assert entry['usd_value'] is not None
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_statistics.py:328]: Please also assert that usd value appears in descending order.

File: rotkehlchen/tests/api/test_statistics.py [hunk 14]
```diff
@@ -303,19 +313,35 @@ def assert_okay_by_location(response):
...
+                'ETH2': get_asset_balance_total(A_ETH2, setup),
+                'BTC': get_asset_balance_total(A_BTC, setup),
+                'EUR': get_asset_balance_total(A_EUR, setup),
+                A_RDN.identifier: get_asset_balance_total(A_RDN, setup),
+            }
+            for entry in result:
+                assert len(entry) == 5
+                assert entry['time'] >= start_time
+                assert entry['category'] == 'asset'
+                assert entry['usd_value'] is not None
```
Comment (LefterisJP) [inline rotkehlchen/tests/api/test_statistics.py:343]: same here

File: tools/data_faker/data_faker/statistics.py [hunk 15]
```diff
@@ -177,8 +177,8 @@ def create_fake_data(self, cursor: DBCursor, args: argparse.Namespace) -> None:
                     category=BalanceType.ASSET,
                     time=from_ts,
                     asset=assets[idx],
-                    amount=str(new_amount),
-                    usd_value=str(action(FVal(assets_data[idx].usd_value), value)),
+                    amount=FVal(new_amount),
```
Comment (LefterisJP) [inline tools/data_faker/data_faker/statistics.py:180]: you don't need this ... you are already dealing with FVals
```suggestion
                    amount=new_amount,
```

File: tools/data_faker/data_faker/statistics.py [hunk 16]
```diff
@@ -177,8 +177,8 @@ def create_fake_data(self, cursor: DBCursor, args: argparse.Namespace) -> None:
...
+                    usd_value=FVal(action(FVal(assets_data[idx].usd_value), value)),
```
Comment (LefterisJP) [inline tools/data_faker/data_faker/statistics.py:181]: Again ... please read the code you are changing. To me it really seems as if you don't understand what you are doing and are just mechanically pasting stuff around.

1. You don't need to cast the whole thing to FVal since Fval + Fval -> Fval and Fval - Fval -> Fval
2. And thanks to your change `usd_value` of a `DBAssetBalance` is now an FVal
```suggestion
                    usd_value=action(assets_data[idx].usd_value, value),
```

**Review by LefterisJP — COMMENTED** [review]
Still lots of things to change.

This should have been ready by now but it really seems to me as if you are not understanding or reading the code you are changing.

**Isaac please read the code you are touching, understand what it does and what changes you should make around it**.

Do no just simply mechanically copy paste stuff around hoping it will work until the few linters we got tell you it's fine.

File: rotkehlchen/db/dbhandler.py [hunk 17]
```diff
@@ -2772,27 +2768,50 @@ def get_latest_asset_value_distribution(self) -> List[DBAssetBalance]:
...
+                            amount=amount,
+                            usd_value=usd_value,
+                            category=BalanceType.deserialize_from_db(result[4]),
+                        ),
+                    )
+            # only add the eth_balance if it contains a balance > 0
+            if eth_balance.amount > ZERO or eth_balance.usd_value > ZERO:
+                # respect descending order `usd_value`
+                for index, balance in enumerate(asset_balances):
+                    if eth_balance.usd_value > balance.usd_value:
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2812]: There is a bug here. If the ETH balance is the smallest of the user's balances it will never be included in the result.

You can use the `for ... else` clause and in the `else part append it at the end of the list.

File: rotkehlchen/db/dbhandler.py [hunk 18]
```diff
@@ -2772,27 +2768,50 @@ def get_latest_asset_value_distribution(self) -> List[DBAssetBalance]:
...
+                        DBAssetBalance(
+                            time=time,
+                            asset=asset,
+                            amount=amount,
+                            usd_value=usd_value,
+                            category=BalanceType.deserialize_from_db(result[4]),
+                        ),
+                    )
+            # only add the eth_balance if it contains a balance > 0
+            if eth_balance.amount > ZERO or eth_balance.usd_value > ZERO:
```
Comment (LefterisJP) [inline rotkehlchen/db/dbhandler.py:2809]: no need for both checks

```suggestion
            if eth_balance.amount > ZERO:
```

File: rotkehlchen/db/utils.py [hunk 19]
```diff
@@ -147,33 +147,65 @@ def get(self, blockchain: SupportedBlockchain) -> ListOfBlockchainAddresses:
...
+        """
         if export_data:
             return {
                 'timestamp': timestamp_to_date(self.time, '%Y-%m-%d %H:%M:%S'),
                 'category': self.category.serialize(),
                 'asset': str(self.asset),
-                'amount': self.amount,
-                f'{export_data[0].symbol.lower()}_value': str(FVal(self.usd_value) * export_data[1]),  # noqa: 501
+                'amount': str(self.amount),
+                f'{export_data[0].symbol.lower()}_value': str(FVal(self.usd_value) * export_data[1]),  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/db/utils.py:172]: `self.usd_value` is already an fval ...

```suggestion
                f'{export_data[0].symbol.lower()}_value': str(self.usd_value * export_data[1]),  # noqa: E501
```

File: rotkehlchen/db/utils.py (hunk 11 above)
Comment (LefterisJP) [reply rotkehlchen/db/utils.py:161]: This is still not fixed. You wrote:

```
If `export_data` is provided, the data is serialized for human consumption i.e
`usd_value` is converted to the user's currency & timestamp is converted to a readable one.
```

I still don't understand what is `export_data`. What are the possible values for it?

**Review by LefterisJP — APPROVED** [review]
LGTM. Thank you Isaac. Try to keep in mind what I told you in the reviews for next PRs.
